app/javascript/lib/scatter-plot.js

```diff
@@ -0,0 +1,308 @@
+ */
+export function resizeScatterPlots() {
+  const plots = window.SCP.scatterPlots
+
+  plots.forEach(rawPlot => {
+    const target = rawPlot.plotId
+    const layout = getScatterPlotLayout(rawPlot)
+    Plotly.relayout(target, layout)
+  })
+}
+
+/** Change Plotly scatter plot color scales */
+export function setScatterPlotColorScales(theme) {
+  const plots = window.SCP.scatterPlots
+
+  plots.forEach(rawPlot => {
+    const target = rawPlot.plotId
+    const dataUpdate = { 'marker.colorscale': theme }
+    Plotly.update(target, dataUpdate)
+  })
+}
+
+/** Get Plotly layout object for scatter plot */
+export function getBaseLayout(height, width) {
+  const font = labelFont
+
+  const layout = {
+    hovermode: 'closest',
+    margin: {
+      t: 25,
+      r: 0,
+      b: 20,
+      l: 0
+    },
+    height,
+    width,
+    font
+  }
+  return layout
+}
+
+/** Gets Plotly layout scene props for 3D scatter plot */
+export function get3DScatterProps(camera, cluster) {
+  const { domainRanges, axes } = cluster
+  const { titles, ranges, aspects } = axes
+
+  const scene = {
+    camera,
+    aspectmode: 'cube',
+    xaxis: { title: titles.x, autorange: true, showticklabels: false },
+    yaxis: { title: titles.y, autorange: true, showticklabels: false },
+    zaxis: { title: titles.z, autorange: true, showticklabels: false }
+  }
+
+  if (domainRanges) {
+    scene.xaxis.autorange = false
+    scene.xaxis.range = ranges.x
+    scene.yaxis.autorange = false
+    scene.yaxis.range = ranges.y
+    scene.zaxis.autorange = false
+    scene.zaxis.range = ranges.x
+    scene.aspectmode = aspects.mode,
+    scene.aspectratio = {
+      x: aspects.x,
+      y: aspects.y,
+      z: aspects.z
+    }
+  }
+
+  return scene
+}
+
+/** Gets Plotly layout props for 2D scatter plot */
+export function get2DScatterProps(cluster) {
+  const {
+    axes, domainRanges, hasCoordinateLabels, coordinateLabels
+  } = cluster
+  const { titles } = axes
+
+  const layout = {
+    xaxis: { title: titles.x, showticklabels: false },
+    yaxis: { title: titles.y, showticklabels: false, scaleanchor: 'x' }
+  }
+
+  // if user has supplied a range, set that, otherwise let Plotly autorange
+  if (domainRanges) {
+    layout.xaxis.range = domainRanges.x
+    layout.yaxis.range = domainRanges.y
+  } else {
+    layout.xaxis.autorange = true
+    layout.yaxis.autorange = true
+  }
+
+  if (hasCoordinateLabels) {
+    layout.annotations = coordinateLabels
+  }
+
+  return layout
+}
+
+/** Set colors for group annotations */
+export function setMarkerColors(data) {
+  return data.map((trace, i) => {
+    trace.marker.color = getColorBrewerColor(i)
+    return trace
+  })
+}
+
+/** Get height and width for a to-be-rendered cluster plot */
+function calculatePlotRect(numRows, numColumns) {
+  // Get height
+  const $ideo = $('#_ideogram')
+  const ideogramHeight = $ideo.length ? $ideo.height() : 0
+  const verticalPad = 225 // Accounts for page header, search area, nav tabs
```

Comment:
hardcoding this (and the horizontalPad below) means that this plot function can't be used on any other page? Maybe make them default params or at least pull them into file-wide constants to make it clearer where the limit is coming from?

Reply:
Good suggestion with default parameters, done.